experimental/algorithm/LAGraph_CFL_reachability.c

```diff
@@ -0,0 +1,324 @@
+    size_t term_rules[rules_count], term_rules_count = 0; // [Variable -> term]
+    size_t bin_rules[rules_count], bin_rules_count = 0;   // [Variable -> AB]
+
+    // Process rules
+    typedef struct {
+        size_t count;
+        size_t len_indexes_str;
+        char indexes_str[LAGRAPH_MSG_LEN];
+    } rule_error_s;
+    rule_error_s term_err = {0};
+    rule_error_s nonterm_err = {0};
+    rule_error_s invalid_err = {0};
+    for (size_t i = 0; i < rules_count; i++) {
+        LAGraph_rule_WCNF rule = rules[i];
+
+        bool is_rule_eps = rule.prod_A == -1 && rule.prod_B == -1;
+        bool is_rule_term = rule.prod_A != -1 && rule.prod_B == -1;
+        bool is_rule_bin = rule.prod_A != -1 && rule.prod_B != -1;
+
+        // Check that all rules are well-formed
+        if (rule.nonterm < 0 || rule.nonterm >= nonterms_count) {
+            ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
+        }
+
+        // [Variable -> eps]
+        if (is_rule_eps) {
+            eps_rules[eps_rules_count++] = i;
+
+            continue;
+        }
+
+        // [Variable -> term]
+        if (is_rule_term) {
+            term_rules[term_rules_count++] = i;
+
+            if (rule.prod_A < -1 || rule.prod_A >= terms_count) {
+                ADD_INDEX_TO_ERROR_RULE(term_err, i);
+            }
+
+            continue;
+        }
+
+        // [Variable -> A B]
+        if (is_rule_bin) {
+            bin_rules[bin_rules_count++] = i;
+
+            if (rule.prod_A < -1 || rule.prod_A >= nonterms_count || rule.prod_B < -1 ||
+                rule.prod_B >= nonterms_count) {
+                ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
+            }
+
+            continue;
+        }
+
+        // [Variable -> _ B]
+        ADD_INDEX_TO_ERROR_RULE(invalid_err, i);
+    }
+
+    if (term_err.count + nonterm_err.count + invalid_err.count > 0) {
+        ADD_TO_MSG("Count of invalid rules: %ld.\n",
+                   term_err.count + nonterm_err.count + invalid_err.count);
+
+        if (nonterm_err.count > 0) {
+            ADD_TO_MSG("Non-terminals must be in range [0, nonterms_count). ");
+            ADD_TO_MSG("Indexes of invalid rules: %s\n", nonterm_err.indexes_str)
+        }
+        if (term_err.count > 0) {
+            ADD_TO_MSG("Terminals must be in range [-1, nonterms_count). ");
+            ADD_TO_MSG("Indexes of invalid rules: %s\n", term_err.indexes_str)
+        }
+        if (invalid_err.count > 0) {
+            ADD_TO_MSG("[Variable -> _ B] type of rule is not acceptable. ");
+            ADD_TO_MSG("Indexes of invalid rules: %s\n", invalid_err.indexes_str)
+        }
+
+        LG_FREE_ALL;
+        return GrB_INVALID_VALUE;
+    }
+
+    // Rule [Variable -> term]
+    for (size_t i = 0; i < term_rules_count; i++) {
+        LAGraph_rule_WCNF term_rule = rules[term_rules[i]];
+
+        GrB_Index nnz;
+        GrB_Matrix_nvals(&nnz, adj_matrices[term_rule.prod_A]);
+
+        GrB_Index row[nnz], col[nnz];
```

Comment:
Is it good idea to use VLA here?

Reply:
Nope

![image](https://github.com/user-attachments/assets/efcfded1-2a3b-4ea1-88be-e6386ae6d204)